Blog on PSP deprecation

kaslin opened the pull request:
The Comms team was requested to write up a blog post to alleviate concerns the community may have about the deprecation of Pod Security Policies. Here's the original request: https://kubernetes.slack.com/archives/C1TU9EB9S/p1611185981053300

It relates to this action taken: kubernetes/kubernetes#97171 (comment)

This pull request consists of the draft of said blog post.

cc @mrbobbytables @mbbroberg @chris-short

Fixes #26202

kaslin commented:
/sig contributor-experience
/area contributor-comms

mbbroberg commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+slug: psp-deprecation
+---
+
+**Authors:** Matt Broberg (RedHat), Kaslin Fields (Google), Chris Short (RedHat)
```
```suggestion
**Authors:** Matt Broberg (Red Hat), Kaslin Fields (Google), Chris Short (Red Hat)
```

mbbroberg commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+To watch an explanation of PSP removal, you can check out this session by SIG-Auth at KubeCon NA 2019:
+
+[https://youtu.be/SFtHRmPuhEw?t=953](https://youtu.be/SFtHRmPuhEw?t=953) 
```
Could we embed the video? Need to defer to a blog team member on whether this will work. 

```suggestion
<iframe width="560" height="315" src="https://www.youtube.com/embed/SFtHRmPuhEw?start=953" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
```

mbbroberg replied:
Ignore this -- we should use the Hugo short code instead of iframes.

mbbroberg commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.
+
+The removal of PSP will pave the way for adoption of more suitable and extensible security policy enforcement options.
```
I get stuck on `suitable` -- on knowing what that means. It might flow better focused on extensibility. 

```suggestion
The removal of PSP will pave the way for adoption of more extensible security policy enforcement options.
```

mbbroberg commented:
/lgtm

Leaving approval to a friend in docs!

mbbroberg commented:
For realz this time 

/lgtm

divya-mohan0209 commented:
👋 everyone! Thank you for all your efforts on this!

My two cents on the draft as the Comms lead for 1.21 Release:

1. We are narrowing down the dates and are still in discussion for the final merge/publish date since alternatives for the deprecated feature are possibly going to be included in this cycle. Would request y'all to hold this till we have a confirmation. cc: @palnabarun @justaugustus 
2. I'm not sure if a tech review has been requested from the sig-auth folks on this PR. @tabbysable would this be something you could look at once you have more bandwidth and there is more clarity around what is going to be included?

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+---
+layout: blog
+title: 'What deprecation of Pod Security Policies means'
```
Comment from SIG Docs weekly meeting:
we recommend having “replace” or “replacement” in the article title, rather than “deprecate” or “deprecation”

kaslin replied:
One thing we struggled with here is that SIG-Auth didn't have a specific suggestion for a replacement technology. (At least from what reviews we got from them.)  So "replacement" wouldn't be quite correct for what we're actually talking about here? We tried to arrange this post to explain that the deprecation process doesn't mean it's going away immediately, and that there are other options, but we don't want to recommend a specific one. Our suggestion really is to take a look at external options and choose the right one on a case-by-case basis.

kaslin replied:
A new commit has been pushed to suggest a new title. Thank you for the feedback and please do check it out!

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritypolicy-v1beta1-policy) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They a ...
+
+Doc: [https://kubernetes.io/docs/concepts/policy/pod-security-policy/](https://kubernetes.io/docs/concepts/policy/pod-security-policy/)
```
```suggestion
The [PodSecurityPolicy](/docs/concepts/policy/pod-security-policy/)
documentation explains this in more detail.
```
?

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+## What do Pod Security Policies do now?
+
+> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritypolicy-v1beta1-policy) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They a ...
```
```suggestion
> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.
```

sftim replied:
Take care when hyperlinking to https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritypolicy-v1beta1-policy as this will _not_ stay current after the v1.21 release. We are switching to a new API reference (that will have stable URIs), see suggestion above.

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+**Authors:** Matt Broberg (Red Hat), Kaslin Fields (Google), Chris Short (Red Hat)
+
+Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](https://kubernetes.io/docs/reference/access-authn-auth ...
```
```suggestion
Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](/docs/reference/access-authn-authz/admission-controllers/). For more details, keep reading.
```

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+## What will replace them? 
+
+Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards ...
```
```suggestion
Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](/pod-security-standards/) page in the Kubernetes documentation aims “to detail recommended Pod security profiles, decoupled from any specific instantiation.”  This page is a useful reference for common pod security configurations, which can be implemented by native or external pod security enforcement tools.
```

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](https://kubernetes.io/docs/concepts/security/pod-security-standards ...
+
+While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Contro ...
```
```suggestion
While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Controller Webhook in Kubernetes, but there are a variety of available options. The blog post, [“A Guide to Kubernetes Admission Controllers”](/blog/2019/03/21/a-guide-to-kubernetes-admission-controllers/) from March 2019 gives a great overview of what admission controller webhooks can do.
```

sftim reviewed (commented):
Markdown & wording feedback

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,48 @@
+
+Deprecation notices draw our attention to the wide range of work going on in the Kubernetes ecosystem that will change the way we operate Kubernetes clusters in the future. This work is taken quite seriously and goes through a detailed exploration process by related SIG leads to understand the implications.
+
+The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.
```
I worry about this message. My concern is that people who skip to the end of the article will infer that the overall message is:
1. PodSecurityPolicy is going
2. There isn't yet a viable alternative

and that this might not go down well with all folks.

kaslin replied:
Thanks! I'll work on reworking this to be more positive and more of a summary. (Whereas it's currently more conclusion than summary.)

justaugustus commented:
I haven't had a chance to dive into this, but explicit hold on this until we have the following groups (at a baseline) check-in:
- Tech review (@kubernetes/sig-auth-pr-reviews @kubernetes/sig-security-leads)
- Overall messaging (@kubernetes/sig-release-leads)

and further that those groups should be mentioned as authors/editors to further validate the path.

/hold

On a quick glance, I'd agree with @sftim's [review](https://github.com/kubernetes/website/pull/26279#discussion_r568873776), that a deprecation without a clear path forward is going to be scary for consumers.

justaugustus commented:
/sig security release

sftim commented:
With https://github.com/kubernetes/website/pull/26279#issuecomment-771933671 in mind
/lgtm cancel

chris-short commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+Deprecation notices draw our attention to the wide range of work going on in the Kubernetes ecosystem that will change the way we operate Kubernetes clusters in the future. This work is taken quite seriously and goes through a detailed exploration process by related SIG leads to understand the implications.
+
+The deprecation of PSP will begin with Kubernetes version 1.21, with its removal in a future version. As we approach that milestone, the Kubernetes community will be working to document viable alternatives.
```
It's slotted to be removed in 1.25 but, that's far enough out for it to slip or move. We didn't necessarily want to lock the community into a hard and fast date/release.

liggitt replied:
+1 for putting a concrete end-release instead of a vague "future version". The removal release is explicitly 1.25, and is surfaced in release notes and in warnings returned by the API server.

kaslin replied:
A new commit has been pushed to address this.

mbbroberg commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+slug: psp-deprecation
+---
+
+**Authors:** Matt Broberg (RedHat), Kaslin Fields (Google), Chris Short (RedHat)
```
I have a suggested edit in there for that above.

chris-short commented:
> I haven't had a chance to dive into this, but explicit hold on this until we have the following groups (at a baseline) check-in:
> 
>     * Tech review (@kubernetes/sig-auth-pr-reviews @kubernetes/sig-security-leads)
> 
>     * Overall messaging (@kubernetes/sig-release-leads)
> 
> 
> and further that those groups should be mentioned as authors/editors to further validate the path.
> 
> /hold
> 
> On a quick glance, I'd agree with @sftim's [review](https://github.com/kubernetes/website/pull/26279#discussion_r568873776), that a deprecation without a clear path forward is going to be scary for consumers.

I think everyone needs to look at what the goal of this blog post is. Something is being deprecated (not many people understand that process in K8s). We're trying to get AHEAD of the programmatic notices and potential FUD that springs forth from them.

Clarity is important. But potentially more so, in this case, is timeliness. We don't have the full picture yet as a community from what I can tell so perhaps we should say we're figuring that out and there'll be more updates to come. But, the goal is to get this piece out the door well before the release.

mbbroberg commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+---
+
+**Authors:** Matt Broberg (Red Hat), Kaslin Fields (Google), Chris Short (Red Hat)
+
```
```suggestion
_The following is a brief overview of upcoming changes to Kubernetes. Bookmark and [keep an eye on our release site](https://www.kubernetes.dev/resources/release/) to see the latest news and other important updates._
```

mbbroberg replied:
^ This is intended to reinforce the scope from the beginning of the article. If there's a better place to link them, let's do it here. cc @justaugustus @divya-mohan0209

divya-mohan0209 replied:
@mbbroberg : I agree with this reinforcement of scope at the very start of the blog. It makes total sense to lay it out & convey that this is not a technical deep-dive and shouldn't be treated as one.

mbbroberg commented:
Hey all, this could benefit from clarifying the scope a bit.

This article is intended to be an editorialized review of the deprecation notice by @kaslin, @chris-short, and me in collaboration and at the request of @mrbobbytables. 

This article is **not** a comprehensive view of all opinions and expertise in our community on the PSP deprecation. 

RE: 

```
explicit hold on this until we have the following groups (at a baseline) check-in:

Tech review (@kubernetes/sig-auth-pr-reviews @kubernetes/sig-security-leads)
Overall messaging (@kubernetes/sig-release-leads)

and further that those groups should be mentioned as authors/editors to further validate the path.
```

Let's do what's necessary to make the scope clear as is. 

Any other k8s contributors and leads can write subsequent blog posts about their technical overview or personal opinions on the best next steps. 

Feedback is welcome within that scope but is otherwise best offered up as another article.

mbbroberg commented:
Oh one additional idea -- if anyone wants to be quoted in the article, that could be a streamlined way to get involved at this stage. Add a statement to this PR and we can work it in.

divya-mohan0209 commented:
> Hey all, this could benefit from clarifying the scope a bit.
> 
> This article is intended to be an editorialized review of the deprecation notice by @kaslin, @chris-short, and me in collaboration and at the request of @mrbobbytables.
> 
> This article is **not** a comprehensive view of all opinions and expertise in our community on the PSP deprecation.
> 
> RE:
> 
> ```
> explicit hold on this until we have the following groups (at a baseline) check-in:
> 
> Tech review (@kubernetes/sig-auth-pr-reviews @kubernetes/sig-security-leads)
> Overall messaging (@kubernetes/sig-release-leads)
> 
> and further that those groups should be mentioned as authors/editors to further validate the path.
> ```
> 
> Let's do what's necessary to make the scope clear as is.
> 
> Any other k8s contributors and leads can write subsequent blog posts about their technical overview or personal opinions on the best next steps.
> 
> Feedback is welcome within that scope but is otherwise best offered up as another article.

To the best of my knowledge, I am sure most of us agree upon the scope. However, there are a couple of things that are uncertain given the timeline we're pushing this out. 

- In the last cycle, the dockershim news broke out pretty close to the release. With a blog announcing the deprecation two months prior, I believe it's also necessary for us to stage the content appropriately so we don't contribute to the very thing we're trying to avoid. We will be following up with a feature blog  (hopefully) covering the technicalities after the release, however one of the important things is to cover all possible initial questions that might arise. 
- Additionally, I believe @justaugustus had cc'ed in the folks from sig-auth & sig-security since we want to ensure we're not missing out anything major + the fact that they're currently working on alternatives to the deprecation. If they do get it in this cycle, we could include that in the blog as well. 
- The suggestion to include the sig names in the byline or having them quoted explicitly is for added reassurance. @sftim would be a better person to weigh in on which is the better option for this bit.

cc: @Pensu @palnabarun

kaslin commented:
Hello everyone! I've updated the title and conclusion and added in the specific version (1.25) when PSPs are expected to be deprecated. I will update the file name to something along the lines of *date*-PSP-Updates.md when a version of the text is approved. These updates better reflect the scope and objectives of this blog post.

@divya-mohan0209, @saschagrunert, @sftim your review would be appreciated.

cc: @justaugustus @Pensu @palnabarun

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+To watch an explanation of PSP removal, you can check out this session by SIG-Auth at KubeCon NA 2019:
+
+{{< youtube "SFtHRmPuhEw?start=953" >}} 
```
Looks massive on my (2160p) screen
```suggestion
{{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}
```
perhaps?

sftim reviewed (commented):
A few more nits, I'm afraid.

sftim commented:
BTW @kaslin you don't need to put the article date in the filename as Hugo already puts it in the URL automatically.
Check out the current [preview](https://deploy-preview-26279--kubernetes-io-master-staging.netlify.app/blog/2021/02/04/psp-update-2021/) and you'll see that the path is `/blog/2021/02/04/psp-update-2021/`.

sftim commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+## What do Pod Security Policies do now?
+
+> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod | specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator t ...
```
Also, tiny tweak:
```suggestion
> A Pod Security Policy is a cluster-level resource that controls security sensitive aspects of the pod specification. The [PodSecurityPolicy](/docs/reference/kubernetes-api/policies-resources/pod-security-policy-v1beta1/) object defines a set of conditions that a pod must run with in order to be accepted into the system, as well as defaults for the related fields. They allow an administrator to control a number of security-related functions.
```

sftim replied:
Could make that italic (wrap it in underscores) if you like.

sftim commented:
@kubernetes/sig-docs-blog-reviewers please also comment on this - your say on the article is important.

kaslin commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+slug: psp-update-2021
+---
+
+**Authors:** Contributor Comms Team
```
Added.

kaslin replied:
Does this need two new lines though?

kaslin commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+
+**Authors:** Contributor Comms Team
+_The following is a brief overview of upcoming changes to Kubernetes. Bookmark and [keep an eye on our release site](https://www.kubernetes.dev/resources/release/) to see the latest news and other important updates._
+Pod Security Policies (PSP), starting with the Kubernetes 1.21, will begin the [process of deprecation](/docs/reference/using-api/deprecation-policy/) with the intention to fully remove it in a future release. This bears the question: what will replace Pod Security Policies in the future? In short, [Admission Controllers](/docs/reference/access-authn-authz/admission-controllers/). For more detail ...
```
Changed to "prompts"

mbbroberg commented:
We had a good catch up call today. Everyone is in agreement that it's best to get a post out there that covers the change, but only if it has a meaningful call to action based on the timeline and desired outcomes. 

We can come to a consensus on whether that looks like. Here are some options that come to mind:  

1. No CTA (fastest timeline, positive impact to awareness, negative impact to certainty)
2. A link to the latest news on the release (fast timeline, positive impact to awareness)
3. A note on the replacement technology (potentially slow timeline, positive impact on certainty & awareness)
4. A recommendation on replacement (slowest timeline, most positive impact on certainty)

This list might help clarify the intended scope. We aimed for 2 with the goal that 3 and 4 would follow from release team efforts as certainty increased. cc @Pensu for the next steps and welcome other's thoughts

divya-mohan0209 commented:
For better clarity @mbbroberg , what that potentially means for us in terms of next steps is this: 

- Get this out in the next couple of weeks with the caveat that it is as updated as possible
- A subsequent note/update to this post or another separate post as and when we have more clarity.
- A technical deep dive of the replacement alternative available (if it goes in, this cycle) as a feature blog, after the release.

Would that be a good set of next steps per what you described above? Or did I miss something?

Cc: @justaugustus @palnabarun

kbhawkey commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+
+## Why are Pod Security Policies being deprecated? 
+
+To watch an explanation of PSP removal, you can check out this session by SIG-Auth at KubeCon NA 2019:
```
nit: "To watch an explanation of PSP removal ..."? Possibly reword to  something like:
To [better] understand the development of PSP ... and why PSP is being deprecated OR
To understand why PSP ...

Since the video is from 2019, I am guessing that your intent is to provide some historical perspective about the upcoming changes. This section describes the basic rationale for removing or deprecating PSP (without going into much technical detail). The video is backing up your claim and reasoning for the taking these steps to change.

kbhawkey commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+To quote the above video: "Various limitations and structural problems have prevented the PSP API from GA." Having objects in Kubernetes never reaching GA (“permabeta”) is a problem for administrators and organizations, as a non-GA feature is not considered to be ready for widespread use. Given the structural problems with the PSPs as they exist today, there is no real path for them to become GA. ...
+
+
+## What will replace them? 
```
nit: replace "them" with Pod Security Policies?
nit: "While the scope of PSPs is beyond anything that will replace it". This paragraph is somewhat vague. Will you indicate that more communication/documentation about ACWs and security policies is coming.

kbhawkey commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+## Wrapping it up
+
+Pod Security Policies were an early attempt to support users' needs with regard to securing pods. Since their creation, they have become unwieldy within the larger context of Kubernetes, making them both difficult to maintain and difficult for users to use.
+
```
nit; Possibly reword, "... for users to use"? Something like:
"making them difficult to configure, manage, and maintain."

tabbysable commented:
/hold huge edit incoming next day or two

mikedanese commented:
/assign @tallclair

rikatz commented:
@kubernetes/sig-docs-pt-reviews fyi so we can keep this on radar for localization

tallclair commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+
+{{< youtube "SFtHRmPuhEw?start=953" youtube-quote-sm >}}
+
+In short, PSPs are part of the kubelet and API server. The tight coupling to these parts of Kubernetes makes PSPs cumbersome both for developers and for operators. Using PSPs is often unintuitive, and setting them up too often leads to unintended results. The initial design of PSPs did not foresee many of the use cases they’ve since tried to solve.
```
I don't actually think there's a problem with the tight coupling to the API server. Both of the replacement designs we're considering are still built-in. [This document](https://docs.google.com/document/d/1VKqjUlpU888OYtIrBwidL43FOLhbmOD5tesYwmjzO4E/edit#heading=h.3qhd5uuytbqs) has a summary of what I see as the top issues with PSP. How about this as an alternative?
```suggestion
In short, there are 3 main challenges with PSPs today:
1. A flawed policy authorization model - Policies can be bound either to the pod's service account, or the requester. This dual binding approach weakens security, and often leads to confusion and breakage.
2. Rollout challenges - PSP is enabled or disabled for the entire cluster, lacks a non-enforcing or dryrun mode, and fails closed. This combination makes it very difficult to enable PSP by default, or rollout to an existing cluster.
3. Inconsistent API - PSP has grown organically over many releases but doesn't have a clear scope of what should be included. Backwards compatibility issues has lead to inconsistencies in the default behavior of several fields, and mutations can have unexpected side effects.
```

tallclair commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+
+Securing your pods requires understanding how the pods are intended to run, and what types of behaviors or activities you seek to prevent. While Pod Security Policies have been one way to enforce security configurations, other tools in the cloud native ecosystem have security enforcement capabilities. The [Pod Security Standards](/docs/concepts/security/pod-security-standards/) page in the Kubern ...
+
+While the scope of PSPs is beyond anything that will replace it, there are [a few tools in the ecosystem](https://landscape.cncf.io/card-mode?category=security-compliance&grouping=category) that can help enforce security policies through the management of Admission Controller Webhooks. We encourage you to assess the right tool for your needs. Pod Security Policies are one type of Admission Contro ...
```
> While the scope of PSPs is beyond anything that will replace it

I'm not sure what this means. Do you mean that none of the replacements we're considering cover the full scope of what PSPs do?

tallclair replied:
> Pod Security Policies are one type of Admission Controller Webhook in Kubernetes

PSPs are a built in controller, not a webhook

tallclair commented on content/en/blog/_posts/2021-01-28-PSP-deprecation.md:
```diff
@@ -0,0 +1,50 @@
+To quote the above video: "Various limitations and structural problems have prevented the PSP API from GA." Having objects in Kubernetes never reaching GA (“permabeta”) is a problem for administrators and organizations, as a non-GA feature is not considered to be ready for widespread use. Given the structural problems with the PSPs as they exist today, there is no real path for them to become GA. ...
+
+
+## What will replace them? 
```
When is this targeted for publication? We are still hoping to have agreed upon a replacement design prior to v1.21. Even if we haven't selected a replacement design, I think it's worth mentioning that there is a sig-auth breakout group working towards a replacement.

sftim replied:
I don't think a publication date is set yet.

chris-short replied:
This should've gone out weeks ago when people started asking me.

IanColdwater replied:
Personally, I'm glad this post didn't go out weeks ago, because SIGs Auth and Security have been making good progress on a solution that was not reflected in this post. At this time, we are planning on closing this PR, rewriting a post to more accurately reflect current events, and putting in a new one. We feel that a clearer post with more definitive solutions will provide better value for users overall than sooner publication of a post without those things would have. Thanks! See you in that PR :)

kaslin commented:
Per @IanColdwater's comment and discussion with @tabbysable, SIG-Auth, and SIG-Security- we have decided to close this PR in favor of creating a new post to better explain PSP deprecation and its upcoming replacement solution.

nicktming commented:
> Per @IanColdwater's comment and discussion with @tabbysable, SIG-Auth, and SIG-Security- we have decided to close this PR in favor of creating a new post to better explain PSP deprecation and its upcoming replacement solution.

@kaslin where is the new post?

sftim commented:
@kaslin thanks for your patience on the review process for this article and for helping to kick off a really important discussion!

IanColdwater commented:
> > Per @IanColdwater's comment and discussion with @tabbysable, SIG-Auth, and SIG-Security- we have decided to close this PR in favor of creating a new post to better explain PSP deprecation and its upcoming replacement solution.
> 
> @kaslin where is the new post?

It's being written and a new PR will be created with it.

sftim commented:
Aside: I looked back at https://kubernetes.io/blog/2020/08/21/moving-forward-from-beta/ and I noticed that that article mentions PodSecurityPolicy as a beta API that needed to change or graduate.

sftim commented:
@kaslin  want to say thanks on the time and effort you put into this PR
We went with https://github.com/kubernetes/website/pull/27369 in the end, and to me the final [article](https://kubernetes.io/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/) is better for it - but SIG Docs definitely appreciates the early engagement here.

It's a shame we didn't have the right message agreed earlier.

tabbysable commented:
+1 what Tim said, @kaslin. I know I said this privately, but here it is for the record too: Thanks a ton for the work you put into helping with PSP deprecation comms.
